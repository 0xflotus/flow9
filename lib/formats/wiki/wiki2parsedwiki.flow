export {
	BuildFullWikiTree(src : string) -> [WikiParsedToken];
	ExtractWikiConstructions(lexeme : WikiParsedToken) -> [WikiConstruction];
	token2parsedToken(lexeme : WikiToken) -> WikiParsedToken;
	WikiParsedToken ::= WParsedHeading1, WParsedHeading2, WParsedBullet1, WParsedBullet2, WParsedBullet3, WParsedNumbered1, WParsedText, WParsedIndentedText, WParsedIndentedFirstLine,
			WNewLine, WNewPage, WPageSeparator, WSmall, WParsedCenter, WParsedTable, WParsedCoach, WParsedRightNumbered, WEOL, WPageName, WComment;
		WParsedHeading1 : (constructions : [WikiConstruction]);
		WParsedHeading2 : (constructions : [WikiConstruction]);
		WParsedBullet1 : (constructions : [WikiConstruction]);
		WParsedBullet2 : (constructions : [WikiConstruction]);
		WParsedBullet3 : (constructions : [WikiConstruction]);
		WParsedNumbered1 : (n : int, pretty : bool, dotted : bool, constructions : [WikiConstruction]);
		WParsedText : (constructions : [WikiConstruction]);
		WParsedIndentedFirstLine : (constructions : [WikiConstruction]);
		WParsedIndentedText : (constructions : [WikiConstruction]);
		WParsedCenter : (token : WikiParsedToken);
		WParsedTable : (cells : [[[WikiConstruction]]]);
		WParsedCoach : (coachConstructions : [WikiConstruction]);
		WParsedRightNumbered : (n : int, token : WikiParsedToken);
}

import formats/wiki/wiki2wikiconstruction;
import formats/wiki/wikimediaextractorshelper;

forbid gui;
forbid renderform;


BuildFullWikiTree(src) {
	lexer = tokenize(src, false);
	for([],
		\r -> lexer.nextlexem() != WEOL(),
		\r -> {
			node = lexer.current() |> token2parsedToken;
			arrayPush(r, node)
		}
	)
}

token2parsedTokenEx(lexeme : WikiToken, insideCenter : bool) -> WikiParsedToken {
	preprocessText = \text -> if (insideCenter) rtrim2(text, " \n\t\u000d") else text;
	makeConstructions = \text -> wiki2wikiConstruction(preprocessText(text), false, [], wikiTableDefaultStyle);
	makeConstructionsStyled = \text, style -> wiki2wikiConstruction(preprocessText(text), false, style, wikiTableDefaultStyle);
	switch (lexeme : WikiToken) {
		// Do not add "default" branch to this switch for easier further search
		WHeading1(text) : text |> makeConstructions |> WParsedHeading1;
		WHeading2(text) : text |> makeConstructions |> WParsedHeading2;
		WBullet1(text) : text |> makeConstructions |> WParsedBullet1;
		WBullet2(text) : text |> makeConstructions |> WParsedBullet2;
		WBullet3(text) : text |> makeConstructions |> WParsedBullet3;
		WNumbered1(n, pretty, dotted, text) : text |> makeConstructions |> \c -> WParsedNumbered1(n, pretty, dotted, c);
		WText(text, multiline) : {
		 	style = if (multiline) [AllowSpacesBeetweenBracketedArgs()] else [];
		 	WParsedText(makeConstructionsStyled(text, style));
		};
		WIndentedFirstLine(text) : text |> makeConstructions |> WParsedIndentedFirstLine;
		WIndentedText(text) : text |> makeConstructions |> WParsedIndentedText;
		WCenter(token) : token2parsedTokenEx(token, true) |> WParsedCenter;
		WTable(cells) : map(cells, \row -> map(row, makeConstructions)) |> WParsedTable;
		WRightNumbered(n, token) : token |> token2parsedToken |> \c -> WParsedRightNumbered(n, c);
		WCoach(coachText) : coachText |> makeConstructions |> WParsedCoach;
		WNewLine() : WNewLine();
		WNewPage(s) : WNewPage(s);
		WPageSeparator(s) : WPageSeparator(s);
		WSmall() : WSmall();
		WPageName(name, condition) : WPageName(name, condition);
		WEOL() : WEOL();
		WComment(text) : WComment(text);
	}
}

token2parsedToken(lexeme : WikiToken) -> WikiParsedToken {
	token2parsedTokenEx(lexeme, false)
}

ExtractWikiConstructions(lexeme) {
	switch(lexeme : WikiParsedToken) {
		// Do not add "default" branch to this switch for easier further search
		WParsedHeading1(constructions) : constructions;
		WParsedHeading2(constructions) : constructions;
		WParsedBullet1(constructions) : constructions;
		WParsedBullet2(constructions) : constructions;
		WParsedBullet3(constructions) : constructions;
		WParsedNumbered1(__, __, __, constructions) : constructions;
		WParsedText(constructions) : constructions;
		WParsedIndentedFirstLine(constructions) : constructions;
		WParsedIndentedText(constructions) : constructions;
		WParsedCenter(token) : ExtractWikiConstructions(token);
		WParsedTable(cells) : mapConcat(cells, concatA);
		WParsedCoach(constructions) : constructions;
		WParsedRightNumbered(__, token) : ExtractWikiConstructions(token);
		WPageName(name, __) : WikiSimpleConstructions([WikiBaseTextEx(name)], name) |> v2a;
		WEOL() : [];
		WNewLine() : [];
		WPageSeparator(__) : [];
		WNewPage(__) : [];
		WSmall() : [];
		WComment(text) : []; //we have to process WikiUnknownMath construction as WComment, but real WComment is not unknown math construction
	}
}
