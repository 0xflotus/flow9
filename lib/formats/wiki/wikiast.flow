import formats/wiki/wikimath;
import formats/wiki/wikivariables;
import paragraphalignments;
import ui/imagedecoratedstructs;
import ui/playeroptions;

export {
	WikiConstruction ::= WikiBaseTextEx, WikiSimpleTextEx, WikiNoWikiEx, WikiWigiEx, WikiAnsEx, WikiSmartArtEx, WikiDecoratedImageEx, WikiOneLineEx, WikiSpanEx,
		WikiTableEx, WikiLearningResourceEx, WikiWriteEx, WikiMediaEx, WikiSubscriptMathEx, WikiSubscriptEx, WikiSuperscriptMathEx,
		WikiSuperscriptEx, WikiSubSupEx, WikiUnderlineEx, WikiBoldItalicEx, WikiBoldEx, WikiItalicEx, ConditionalWikiEx, WikiDebugEx, WikiButtonEx,
		WikiPicButtonEx, WikiSliderEx, WikiScoringIconEx, MathRootEx, MathLongDivisionEx, MathStrikeEx, WikiUnderline2Ex, MathNaturalSyntaxEx,
		WikiRotateEx, WikiScaleEx, WikiMutableEx, MathFracEx, MathBracketEx, DefiniteControlEx, MathScriptEx, MathDottedEx, CustomSymbol,
		WikiTopModificatorEx, WikiDraggableEx, WikiForcedExpandEx, WikiPopupEx, WikiSelectEx, WikiSequenceEx, GlueFragments, WikiNewLine, MathNumber2StringEx, MathNumber2SciNotEx,
		WikiLimiterEx, WikiWhenEx, WikiScoreEx, WikiCaseEx, WikiConditionalAssignment, WikiRepeatEx, WikiComboBoxEx, WikiTimelineEx, WikiBlankEx,
		WikiMessagesCollectorEx, WikiDroptargetEx, WikiPredefinedColorEx, WikiSimpleConstructions, WikiRadioButtonsEx, WikiInlineBullet, WikiCalc, WikiAltTextCollectorEx, MathUnknownConstruction,
		MathAssignmentsOnBecameVisibleEx;

		WikiBaseTextEx(text : string);
		WikiSimpleTextEx(text : string, styleAddition : [CharacterStyle]);
		WikiNoWikiEx(text : string);
		WikiWigiEx(src : string);
		WikiAnsEx(src : string);
		WikiSmartArtEx(src : string);
		WikiDecoratedImageEx(imageDecorated : ImageDecorated);
		WikiOneLineEx(constructions : [WikiConstruction]);
		WikiSpanEx(constructions : [WikiConstruction], tooltip : Maybe<string>, color : Maybe<int>, background : Maybe<int>, fontFamily : Maybe<string>, fontSizeMultiplier : Maybe<Pair<double, bool>>);
		WikiTableEx(constructions : [WikiConstruction], tableStyle : WikiTableStyle);
		WikiLearningResourceEx(lr : string, introduction : Maybe<string>);
		WikiWriteEx(src : string);
		WikiMediaEx(contentType : MediaContent2);
		WikiSubscriptMathEx(constructions : [WikiConstruction]);
		WikiSubscriptEx(src : string);
		WikiSuperscriptMathEx(constructions : [WikiConstruction]);
		WikiSuperscriptEx(src : string);
		WikiSubSupEx(bodyC : [WikiConstruction], supC : [WikiConstruction], subC : [WikiConstruction], left : bool, src : string);
		WikiUnderlineEx(src : string);
		WikiBoldItalicEx(constructions : [WikiConstruction]);
		WikiBoldEx(constructions : [WikiConstruction]);
		WikiItalicEx(constructions : [WikiConstruction], fix : bool);
		WikiConditionalAssignment : (condition : WV_Sequence, thenSequence : WV_Sequence, elseSequence : WV_Sequence);
		MathAssignmentsOnBecameVisibleEx : (sequence : WV_Sequence);
		ConditionalWikiEx(condition : WV_Sequence, thenConstructions : [WikiConstruction], elseConstructions : [WikiConstruction]);
		WikiWhenEx(conditionSequence : WV_Sequence, thenConstructions : [WikiConstruction], elseConstructions : [WikiConstruction]);
		WikiScoreEx(almostCorrectSequence : Maybe<WV_Sequence>, missingSequence : Maybe<WV_Sequence>, conditionSequence : WV_Sequence, tooltip : string, vertical : bool, inIconTooltip : string);
		WikiDebugEx(src : string);
		WikiButtonEx(header : string, seq : WV_Sequence, width : double);
		WikiPicButtonEx(seq : WV_Sequence, button : [WikiConstruction], hovered : Maybe<[WikiConstruction]>, pressed : Maybe<[WikiConstruction]>);
		WikiSliderEx(varName : string, minSeq : WV_Sequence, maxSeq : WV_Sequence, stepSeq : WV_Sequence, widthSeq : WV_Sequence);
		WikiForcedExpandEx(src : string);
		WikiSelectEx(triggerVars : [string], constructions : [WikiConstruction]);
		WikiCaseEx(conditionSeq : WV_Sequence, cases : [[WikiConstruction]]);
		WikiRepeatEx(timesSeq : WV_Sequence, constructions : [WikiConstruction]);
		WikiBlankEx(width : double, matching : WikiBlankMatchingOptions, answers : WikiBlankAnswersData);
			WikiBlankMatchingOptions(caseSensitive : bool, exactMatch : bool, signTolerance : string, decimalTolerance: int);
			WikiBlankAnswersData ::= WikiBlankAnswersLegacyData, WikiBlankAnswers2Data;
				WikiBlankAnswersLegacyData(scoring : WikiBlankLegacyScoring, answers : [WikiBlankAnswer]); //WikiBlankAnswer.type is ignored
				WikiBlankAnswers2Data(scoring : WikiBlank2Scoring, answers : [WikiBlankAnswer]);
				WikiBlankScoring ::= WikiBlankLegacyScoring, WikiBlank2Scoring;
					WikiBlankLegacyScoring(varScore : Maybe<string>, varNAnswer : Maybe<string>, varNumVal : Maybe<string>);
					WikiBlank2Scoring(vGeneralCorrectnessMB : Maybe<string>, vAnswerIndexMB : Maybe<string>, vIsMisstypedMB : Maybe<string>, vAnswerTypeMB : Maybe<string>, vAnswerValueMB : Maybe<string>);
					WikiBlankAnswer(answer : string, type : WikiBlankAnswerType);
		WikiScoringIconEx(type : WikiScoringIconType, tooltip : string);
		WikiComboBoxEx(varName : string, suggestedWidth : double, reorderElements : bool, items : [string]);
		WikiTimelineEx(varName : string, duration : WV_Sequence, maxVal : WV_Sequence, widthVal : WV_Sequence, doLoop : bool, maxLoopsCount : int);
		MathNumber2StringEx(seq : WV_Sequence);
		MathNumber2SciNotEx(numberSequence : WV_Sequence, digitsCountSequence : WV_Sequence, trailingZeroes : bool);
		MathRootEx(degreeConstructions : [WikiConstruction], bodyConstructions : [WikiConstruction], src : string);
		MathLongDivisionEx(bodyConstructions : [WikiConstruction], src : string);
		MathStrikeEx(bodyConstructions : [WikiConstruction], dxSeq : WV_Sequence, dySeq : WV_Sequence, mcolor : Maybe<int>, src : string);
		WikiUnderline2Ex(bodyConstructions : [WikiConstruction], dySeq : WV_Sequence, dlSeq : WV_Sequence, drSeq : WV_Sequence);
		MathNaturalSyntaxEx(src : string);
		WikiRotateEx(bodyConstructions : [WikiConstruction], ansleSeq : WV_Sequence);
		WikiScaleEx(bodyConstructions : [WikiConstruction], scaleSeq : WV_Sequence);
		WikiMutableEx(name : string);
		MathFracEx(enumeratorConstructions : [WikiConstruction], denumeratorConstructions : [WikiConstruction], src : string);
		MathBracketEx(leftBracket : WriteBracketType, rightBracket : WriteBracketType, bodyConstructions : [WikiConstruction], src : string);
		DefiniteControlEx(signType : DefiniteControlSignType, tc : [WikiConstruction], fc : [WikiConstruction], bc : [WikiConstruction]);
		MathScriptEx(type : MathScriptType, bodyConstructions : [WikiConstruction], scriptConstructions : [WikiConstruction], src : string);
			MathScriptType ::= MathOverscript, MathUnderscript, MathSuperscript, MathSubscript;
		MathDottedEx(bodyConstructions : [WikiConstruction], left : int, top : int, right : int, bottom : int, leftSing : bool, rightSign : bool, src : string);
		WikiTopModificatorEx(type : WriteTopModificator, size : WriteTopModificatorSize, bodyConstructions : [WikiConstruction], dySeq : WV_Sequence, dlSeq : WV_Sequence, dxSeq : WV_Sequence, naturalMath : bool);
		WikiDraggableEx(wikiConstructions : [WikiConstruction], shapeConstructions : [WikiConstruction], feedbackConstructions : [WikiConstruction], strokeShapeS : string, fillShapeS : string, idSeq : WV_Sequence, dwSeq : WV_Sequence, dhSeq : WV_Sequence, strkSeq : WV_Sequence, flSeq : WV_Sequence);
		WikiDroptargetEx(varName : string, wikiConstructions : [WikiConstruction], feedbackConstructions : [WikiConstruction], strokeShapeS : string, fillShapeS : string, idSeq : WV_Sequence, dwSeq : WV_Sequence, dhSeq : WV_Sequence, strkSeq : WV_Sequence, flSeq : WV_Sequence);
		WikiPopupEx(wikiConstructions : [WikiConstruction], wdSeq : WV_Sequence, htSeq : WV_Sequence, hASeq : WV_Sequence, vASeq : WV_Sequence);
		WikiSequenceEx(sequence : WV_Sequence);
		WikiLimiterEx(triggrVars : [string], resVar : string, cond : string, valSeq : WV_Sequence);
		WikiMessagesCollectorEx(otherAnswer : string, varName : string, correctAnswer : [string]);
		WikiPredefinedColorEx(color : int, constructions : [WikiConstruction]);
		// Artifical construction for translation purposes
		WikiSimpleConstructions(constructions : [WikiConstruction], src : string);
		WikiRadioButtonsEx(varName : string, wrap : bool, vertical : bool, widthSeq : WV_Sequence, variants : [[WikiConstruction]], variantsS : [string]);
		WikiAltTextCollectorEx(altText : string);



		// Style for table
		WikiTableStyle(graphic : [GraphicsStyle], character : [CharacterStyle], xAlign : double, yAlign : double, xAligns : [double], yAligns : [double], vLines : [int], hLines : [int], left : WriteBracketType, right : WriteBracketType, borders : WikiTableCellBorders);
		WikiTableCellBorders : (left : double, top : double, right : double, bottom : double);
		wikiTableDefaultStyle = WikiTableStyle([], [], 0.0, 0.0, [], [], [], [], WriteNoBracket(), WriteNoBracket(), WikiTableCellBorders(6.0, 6.0, 6.0, 6.0));

		MediaContent ::= VideoContent, PictureContent, ExternalLink, SoundContent, ExpressContent, NotContent;
		MediaContent2 ::= VideoContent, PictureContent, ExternalLink, SoundContent, ExpressContent;
			VideoContent:(url:string, options: [PlayerOption], controls: [PlayerControl], zoom : Maybe<double>);
			PictureContent:(url:string, title : Maybe<string>, zoom: ZoomDescription);
			ExternalLink:(url : string, caption : Maybe<string>);
			SoundContent:(url : string, options : [PlayerOption]);
			ExpressContent:(url : string, title : Maybe<string>, zoom: ZoomDescription);
			NotContent:();

	isSimpleWikiConstruction(construction : WikiConstruction) -> bool;
}


dummyWikiAstAssignment = 0;

isSimpleWikiConstruction(construction : WikiConstruction) {
	checkConstructions = \constructions -> all(map(constructions, isSimpleWikiConstruction));
	switch(construction : WikiConstruction) {
		WikiBaseTextEx(__) : true;
		WikiSimpleTextEx(__, __) : true;
		WikiNoWikiEx(__) : true;
		WikiWigiEx(__) : false;
		WikiAnsEx(__) : false;
		WikiSmartArtEx(__) : false;
		WikiDecoratedImageEx(__) : false;
		WikiOneLineEx(constructions) : checkConstructions(constructions);
		WikiSpanEx(constructions, __, __, __, __, __) : checkConstructions(constructions);
		WikiTableEx(constructions, __) : length(constructions) > 0 && checkConstructions(constructions);
		WikiLearningResourceEx(__, __) : false;
		WikiWriteEx(__) : false;
		WikiMediaEx(__) : false;
		WikiSubscriptMathEx(constructions) : checkConstructions(constructions);
		WikiSubscriptEx(src) : true;
		WikiSuperscriptMathEx(constructions) : checkConstructions(constructions);
		WikiSuperscriptEx(src) : true;
		WikiSubSupEx(bodyC, supC, subC, __, __) : checkConstructions(bodyC) && checkConstructions(supC) && checkConstructions(subC);
		WikiUnderlineEx(src) : true;
		WikiBoldItalicEx(constructions) : checkConstructions(constructions);
		WikiBoldEx(constructions) : checkConstructions(constructions);
		WikiItalicEx(constructions, __) : checkConstructions(constructions);
		WikiConditionalAssignment(__, __, __) : false;
		MathAssignmentsOnBecameVisibleEx(__) : false;
		ConditionalWikiEx(__, thenConstructions, elseConstructions) : checkConstructions(thenConstructions) && checkConstructions(elseConstructions);
		WikiWhenEx(__, thenConstructions, elseConstructions) : checkConstructions(thenConstructions) && checkConstructions(elseConstructions);
		WikiScoreEx(__, __, __, __, __, __) : false;
		WikiDebugEx(__) : true;
		WikiButtonEx(__, __, __) : false;
		WikiPicButtonEx(__, button, hovered, pressed) : checkConstructions(button) && eitherMap(hovered, checkConstructions, true) && eitherMap(pressed, checkConstructions, true);
		WikiSliderEx(__, __, __, __, __) : false;
		WikiForcedExpandEx(src) : false;
		WikiSelectEx(__, constructions) :  checkConstructions(constructions);
		WikiCaseEx(__, cases) : all(map(cases,  checkConstructions));
		WikiRepeatEx(__, constructions) :  checkConstructions(constructions);
		WikiBlankEx(__, __, __) : false;
		WikiScoringIconEx(__, __) : false;
		WikiInlineBullet() : false;
		WikiComboBoxEx(__, __, __, __) : false;
		WikiRadioButtonsEx(__, __, __, __, __, __) : false;
		WikiTimelineEx(__, __, __, __, __, __) : false;
		MathNumber2StringEx(__) : false;
		MathNumber2SciNotEx(__, __, __) : false;
		MathRootEx(degreeConstructions, bodyConstructions, __) : checkConstructions(degreeConstructions) && checkConstructions(bodyConstructions);
		MathLongDivisionEx(bodyConstructions, __) : checkConstructions(bodyConstructions);
		MathStrikeEx(bodyConstructions, __, __, __, __) : checkConstructions(bodyConstructions);
		WikiUnderline2Ex(bodyConstructions, __, __, __) : checkConstructions(bodyConstructions);
		MathNaturalSyntaxEx(src) : false;
		WikiRotateEx(bodyConstructions, __) : checkConstructions(bodyConstructions);
		WikiScaleEx(bodyConstructions, __) : checkConstructions(bodyConstructions);
		WikiMutableEx(__) : false;
		MathFracEx(enumeratorConstructions, denumeratorConstructions, __) : checkConstructions(enumeratorConstructions) && checkConstructions(denumeratorConstructions);
		MathBracketEx(__, __, bodyConstructions, __) : checkConstructions(bodyConstructions);
		DefiniteControlEx(__, tc, fc, bc) : checkConstructions(tc) && checkConstructions(fc) && checkConstructions(bc);
		MathScriptEx(__, bodyConstructions, scriptConstructions, __) : checkConstructions(bodyConstructions) && checkConstructions(scriptConstructions);
		MathDottedEx(bodyConstructions, __, __, __, __, __, __, __) : checkConstructions(bodyConstructions);
		WikiTopModificatorEx(__, __, bodyConstructions, __, __, __, __) : checkConstructions(bodyConstructions);
		WikiDraggableEx(__, __, __, __, __, __, __, __, __, __) : false;
		WikiDroptargetEx(__, __, __, __, __, __, __, __, __, __) : false;
		WikiPopupEx(__, __, __, __, __) : false;
		WikiSequenceEx(__) : false;
		WikiLimiterEx(__, __, __, __) : false;
		WikiMessagesCollectorEx(__, __, __) : false;
		WikiAltTextCollectorEx(__) : false;
		WikiPredefinedColorEx(__, constructions) : checkConstructions(constructions);
		// Artifical construction for translation purposes
		WikiSimpleConstructions(constructions, __) : true;
		GlueFragments() : true;
		WikiNewLine() : false;
		WikiCalc() : false;
		CustomSymbol(__) : true;
		MathUnknownConstruction(__) : true;
	}
}
