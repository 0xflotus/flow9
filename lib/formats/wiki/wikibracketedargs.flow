import array;
import maybe;
import string;

export {
	getArgsList(source : string, expectedArgs : WikiArgsDescription, allowSpaces : bool) -> Maybe<WikiArgsList>;
		WikiArgsDescription : (args : [BracketedArgType]);
			BracketedArgType ::= OptionalArgFn, RequiredArgFn, OptionalArgsListFn;
		WikiArgsList : (args : [WikiArgValue], end : int);
			WikiArgValue ::= SingleArgValue, ListArgValue;
				SingleArgValue : (text : string);
				ListArgValue : (texts : [string]);

		RequiredArgFn(fn : Maybe<(string) -> Maybe<ArgData>>);
		OptionalArgFn(fn : Maybe<(string) -> Maybe<ArgData>>, def : string);
		OptionalArgsListFn(fn : Maybe<(string) -> Maybe<ArgData>>);
			ArgData(text : string, end : int);

		RequiredArg() -> RequiredArgFn;
		OptionalArg(def : string) -> OptionalArgFn;
		OptionalArgsList() -> OptionalArgsListFn;

		RequiredArgDigit() -> RequiredArgFn;
		OptionalArgBracket() -> OptionalArgFn;



	verifyExpectedArgs(expectedArgs : WikiArgsDescription) -> bool;
}

RequiredArg() {
	RequiredArgFn(None())
}
OptionalArg(def) {
	OptionalArgFn(None(), def)
}
OptionalArgsList() {
	OptionalArgsListFn(None())
}

RequiredArgDigit() {
	getDigit = \text -> {
		char = getCharAt(text, 0);
		if (isDigit(char)) {
			Some(ArgData(char, 1))
		} else {
			None()
		}
	}
	RequiredArgFn(Some(getDigit))
}

OptionalArgBracket() {
	getBracket = \text -> {
		char = getCharAt(text, 0);
		if (char == "\\") {
			br = getCharAt(text, 1);
			if (br != "" && strIndexOf("|{}()[]", br) != -1) {
				Some(ArgData(br, 2))
			} else {
				Some(ArgData("", 1))
			}
		} else {
			None()
		}
	}
	OptionalArgFn(Some(getBracket), "")
}

verifyExpectedArgs(expectedArgs) {
	verifyExpectedArgsEx(expectedArgs.args, RequiredArg())
}

verifyExpectedArgsEx(expectedArgs : [BracketedArgType], previousArg : BracketedArgType) {
	len = length(expectedArgs);
	if (len == 0) {
		true
	} else {
		currentArg = expectedArgs[0];
		switch (previousArg) {
		 	RequiredArgFn(__) : verifyExpectedArgsEx(removeIndex(expectedArgs, 0), currentArg);
		 	OptionalArgFn(__, __) : {
		 		switch (currentArg) {
		 			RequiredArgFn(__) : verifyExpectedArgsEx(removeIndex(expectedArgs, 0), currentArg);
		 			OptionalArgFn(__, __) : verifyExpectedArgsEx(removeIndex(expectedArgs, 0), currentArg);
		 			OptionalArgsListFn(__) : {
		 				println("OptionalArg is followed by OptionalArgsList");
		 				false
		 			}
		 		}
		 	}
		 	OptionalArgsListFn(__) : {
		 		switch (currentArg) {
		 			RequiredArgFn(__) : verifyExpectedArgsEx(removeIndex(expectedArgs, 0), currentArg);
		 			OptionalArgFn(__, __) : {
		 				println("OptionalArgsList is followed by OptionalArg");
		 				false
		 			}
		 			OptionalArgsListFn(__) : {
		 				println("OptionalArgsList is followed by OptionalArgsList");
		 				false
		 			}
		 		}
		 	}
		}
	}
}

getArgsListEx(source : string, expectedArgs : [BracketedArgType], argsList : WikiArgsList, allowSpaces : bool) {
	spacesCount = if (allowSpaces) countUntil(0, strlen(source), \i -> isNotSpace(getCharAt(source, i))) else 0;

	checkForFollowingBrackets = \ob, cb -> {
		if (getCharAt(source, spacesCount) == ob)  {
			findMatchingBracketM2(source, ob, cb, "`"); //35807 Support of natural math with brackets inside arguments
		} else {
			None()
		}
	}

	getTextInsideBrackets = \makeDef -> \brackets -> {
		eitherMap(
			brackets,
			\br : Pair<int, int> -> {
				text = substring(source, br.first + 1, br.second - br.first - 2);
				if (isNotSpace(text)) {
					Some(ArgData(text, br.second));
				} else {
					maybeMap(makeDef, \md -> md(br.second))
				}
			},
			None()
		)
	}

	getOptionalParameter = \def -> \__ -> checkForFollowingBrackets("[", "]") |> getTextInsideBrackets(def);
	getRequiredParameter = \__ -> checkForFollowingBrackets("{", "}") |> getTextInsideBrackets(None());


	argsLen = length(expectedArgs);

	if (argsLen == 0) {
		argsList
	} else {
		currentArg = expectedArgs[0];
		switch (currentArg) {
			RequiredArgFn(fn) : {
				eitherMap(
					either(fn, getRequiredParameter)(source),
					\argData -> {
						getArgsListEx(
							strRight(source, argData.end),
							removeIndex(expectedArgs, 0),
							WikiArgsList(
								arrayPush(argsList.args, SingleArgValue(argData.text)),
								argsList.end + argData.end
							),
							allowSpaces
						)
					},
					argsList
				)
			}
			OptionalArgFn(fn, std) : {
				eitherFn(
					either(fn, getOptionalParameter(Some(\end -> ArgData(std, end))))(source),
					\argData -> {
						getArgsListEx(
							strRight(source, argData.end),
							removeIndex(expectedArgs, 0),
							WikiArgsList(
								arrayPush(argsList.args, SingleArgValue(argData.text)),
								argsList.end + argData.end
							),
							allowSpaces
						)
					},
					\ -> {
						getArgsListEx(
							source,
							removeIndex(expectedArgs, 0),
							WikiArgsList(
								arrayPush(argsList.args, SingleArgValue(std)),
								argsList.end
							),
							allowSpaces
						)
					}
				)
			}
			OptionalArgsListFn(fn) : {
				eitherFn(
					either(fn, getOptionalParameter(None()))(source),
					\argData -> {
						newArgs = {
							args = argsList.args;
							len = length(args);
							if (len > 0) {
								switch (args[len - 1]) {
									SingleArgValue(__) : {
										arrayPush(args, ListArgValue([argData.text]))
									}
									ListArgValue(texts) : {
										replace(args, len - 1, ListArgValue(arrayPush(texts, argData.text)))
									}
								}
							} else {
								[ListArgValue([argData.text])]
							}
						};
						getArgsListEx(
							strRight(source, argData.end),
							expectedArgs,
							WikiArgsList(
								newArgs,
								argsList.end + argData.end
							),
							allowSpaces
						)
					},
					\ -> getArgsListEx(source, removeIndex(expectedArgs, 0), argsList, allowSpaces)
				)
			}
		}
	}

}

getArgsList(source, expectedArgs, allowSpaces) {
	if (verifyExpectedArgs(expectedArgs)) {
		argsList = getArgsListEx(source, expectedArgs.args, WikiArgsList([], 0), allowSpaces);
		if (length(expectedArgs.args) == length(argsList.args)) {
			Some(argsList)
		} else {
			None()
		}
	} else {
		None()
	}
}

/*
main() {
	source = "[a][b][c]{d}[e][f][g]";
	expectedArgs = WikiArgsDescription([OptionalArg("defA"), OptionalArg("defB"), OptionalArg("defC"), OptionalArg("defExpected"), RequiredArg(), OptionalArgsList()]);
	getArgsList(source, expectedArgs) |> println
}

*/


