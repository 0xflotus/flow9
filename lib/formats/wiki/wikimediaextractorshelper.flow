export {
	SetSmartCollectMedia : ((string) ->  [string]) -> void;
	CallSmartArtCollectMedia : (string) ->  [string];

	SetWigiMediaExtractor : ((string) -> Maybe<[string]>) -> void;
	CallWigiMediaExtractor : (string) -> Maybe<[string]>;
	SetWigiMediaBlocksExtractor : ((string) -> Maybe<[WigiBlockType]>) -> void;
	CallWigiMediaBlocksExtractor : (string) -> Maybe<[WigiBlockType]>;

	SetSmart2XLIFFNodes : ((int, string) ->  [XmlNode]) -> void;
	CallSmartArt2XLIFFNodes : (int, string) ->  [XmlNode];
}

import formats/xml;
import maybe;
import wigi/types; // WigiBlockType

SmartArtCollectMediaForWiki : ref (string) ->  [string] = ref \s -> [];

SetSmartCollectMedia(parser){
	SmartArtCollectMediaForWiki := parser;
};

CallSmartArtCollectMedia(source) {
	^SmartArtCollectMediaForWiki(source)
}

wigiMediaExtractor : ref (string) -> Maybe<[string]> = ref \__ -> None();

SetWigiMediaExtractor(extractor) {
	wigiMediaExtractor := extractor
}

CallWigiMediaExtractor(wigiElement) {
	^wigiMediaExtractor(wigiElement)
}

wigiMediaBlocksExtractor : ref (string) -> Maybe<[WigiBlockType]> = ref \__ -> None();

SetWigiMediaBlocksExtractor(extractor) {
	wigiMediaBlocksExtractor := extractor
}

CallWigiMediaBlocksExtractor(wigiElement) {
	^wigiMediaBlocksExtractor(wigiElement)
}

SmartArt2XLIFFNodesForWiki : ref (int, string) -> [XmlNode] = ref \i,s -> [];

SetSmart2XLIFFNodes(parser){
	SmartArt2XLIFFNodesForWiki := parser;
};

CallSmartArt2XLIFFNodes(index, source) {
	^SmartArt2XLIFFNodesForWiki(index, source)
}