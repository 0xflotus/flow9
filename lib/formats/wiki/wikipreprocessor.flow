export {
	IuWikiPreprocessor : IuPreprocessor;
	wikiPreprocessDouble(source : string) -> Pair<bool, double>;
}
import illusion/illusion_preprocessor;
import formats/wiki/wikivariables;

DefaultPreprocessDouble(d : IuDoubleVar) -> Behaviour<Pair<bool, double>> {
	select(d.source, wikiPreprocessDouble);
}

wikiPreprocessDouble(source) {
	pr : WV_ParseResult = parseWikiVariables(source);
	switch(pr) {
		WV_CorrectResult(sequence, remainder) : {
			if (isNotSpace(remainder.s)) {
				Pair(false, 0.0)
			} else {
				Pair(true, sequence |> processWVSequence |> s2d)
			}
		}
		WV_RemainderOnly(remainder) : {
			Pair(false, 0.0)
		}
	}
}

IuWikiPreprocessor = IuPreprocessor(makeIntPreprocessor(DefaultPreprocessDouble), DefaultPreprocessDouble);
