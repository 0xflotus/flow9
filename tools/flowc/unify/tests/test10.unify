import wiki;

α5 =() -> α10;
α10 = {WEOL<> .. *};
α1 =ref α8;

() -> WEOL<>  c=  α8;


ref α5 c= . α1;

finalize