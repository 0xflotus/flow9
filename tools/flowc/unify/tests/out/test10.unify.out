α1=ref () -> {WEOL<> .. *}
α5=() -> {WEOL<> .. *}
α8=() -> {WEOL<> .. *}
α10={WEOL<> .. *}
α100={WEOL<> .. *}

Final:
α1=ref () -> WEOL
α5=() -> WEOL
α8=() -> WEOL
α10=WEOL
α100=WEOL
